coverage/jsonreport.py

```diff
@@ -0,0 +1,104 @@
+import sys
+import time
+
+from coverage import __version__
+from coverage.report import get_analysis_to_report
+from coverage.results import Numbers
+
+
+class JsonReporter(object):
+    """A reporter for writing JSON coverage results."""
+
+    def __init__(self, coverage):
+        self.coverage = coverage
+        self.config = self.coverage.config
+        self.total = Numbers()
+        self.report_data = {}
+        self.data = self.coverage.get_data()
+        self.has_arcs = self.data.has_arcs()
+
+    def report(self, morfs, outfile=None):
+        """Generate a json report for `morfs`.
+
+        `morfs` is a list of modules or file names.
+
+        `outfile` is a file object to write the json to
+
+        """
+        outfile = outfile or sys.stdout
+        self.data.set_query_contexts(self.config.report_contexts)
+        meta = {
+            "version": __version__,
+            "timestamp": str(int(time.time())),
+            "branch_coverage": self.data.has_arcs(),
+            "show_contexts": self.config.json_show_contexts,
+        }
+        self.report_data["meta"] = meta
+        measured_files = {
+            file_reporter.relative_filename():
+            self.report_one_file(file_reporter, analysis)
+            for file_reporter, analysis in get_analysis_to_report(
+                self.coverage,
+                morfs
+            )
+        }
+        self.report_data["files"] = measured_files
+
+        self.report_data["totals"] = {
+            'covered_lines': self.total.n_executed,
+            'num_statements': self.total.n_statements,
+            'percent_covered': self.total.pc_covered,
+            'missing_lines': self.total.n_missing,
+        }
+
+        if self.data.has_arcs():
+            self.report_data["totals"].update({
+                'num_branches': self.total.n_branches,
+                'num_partial_branches': self.total.n_partial_branches,
+            })
+
+        outfile.write(
+            json.dumps(
+                self.report_data,
+                indent=4 if self.config.json_pretty_print else None
+            )
+        )
+
+        return self.total.n_statements and self.total.pc_covered
+
+    def report_one_file(self, file_reporter, analysis):
+        nums = analysis.numbers
+        self.total += nums
+        summary = {
+            'covered_lines': nums.n_executed,
+            'num_statements': nums.n_statements,
+            'percent_covered': nums.pc_covered,
+            'missing_lines': nums.n_missing,
+        }
+        reported_file = {
+            'executed_lines': list(sorted(analysis.statements)),
+            'summary': summary,
+            'missing_lines': list(sorted(analysis.missing))
+        }
+        if self.config.json_show_contexts:
+            reported_file['contexts'] = analysis.data.contexts_by_lineno(
+                file_reporter.filename
+            )
+        if self.data.has_arcs():
+            reported_file['summary'].update({
```

Comment:
I like that the JSON leaves out keys that don't apply.  I wonder if people will feel differently! :)

Reply:
Yeah, the tradeoff is parsers will need to be defensive (or know their settings). However, it's a big win for human readability. The simple case will look simple

You may have noticed by now the whole time writing this my head was going back and forth between the human reader and the machine reader.